nebula_ros/src/hesai/decoder_wrapper.cpp

```diff
@@ -0,0 +1,317 @@
+      }
+      else
+      {
+        RCLCPP_INFO_STREAM(logger_, "Saved downloaded data to " << calibration_file_path_from_sensor);
+      }
+    }
+    catch (std::runtime_error& e)
+    {
+      RCLCPP_ERROR_STREAM(logger_, "Could not download calibration data: " << e.what());
+    }
+  }
+
+  // If saved calibration data from a sensor exists (either just downloaded above, or previously), try to load it
+  if (!ignore_others && std::filesystem::exists(calibration_file_path_from_sensor))
+  {
+    auto status = calib->LoadFromFile(calibration_file_path_from_sensor);
+    if (status == Status::OK)
+    {
+      calib->calibration_file = calibration_file_path_from_sensor;
+      return calib;
+    }
+
+    RCLCPP_ERROR_STREAM(logger_, "Could not load downloaded calibration data: " << status);
+  }
+  else if (!ignore_others)
+  {
+    RCLCPP_ERROR(logger_, "No downloaded calibration data found.");
+  }
+
+  if (!ignore_others)
+  {
+    RCLCPP_WARN(logger_, "Falling back to generic calibration file.");
+  }
+
+  // If downloaded data did not exist or could not be loaded, fall back to a generic file.
+  // If that file does not exist either, return an error code
+  if (!std::filesystem::exists(calibration_file_path))
+  {
+    RCLCPP_ERROR(logger_, "No calibration data found.");
+    return nebula::Status(Status::INVALID_CALIBRATION_FILE);
+  }
+
+  // Try to load the existing fallback calibration file. Return an error if this fails
+  auto status = calib->LoadFromFile(calibration_file_path);
+  if (status != Status::OK)
+  {
+    RCLCPP_ERROR_STREAM(logger_, "Could not load calibration file at '" << calibration_file_path << "'");
+    return status;
+  }
+
+  // Return the fallback calibration file
+  calib->calibration_file = calibration_file_path;
+  return calib;
+}
+
+void HesaiDecoderWrapper::ProcessCloudPacket(std::unique_ptr<nebula_msgs::msg::NebulaPacket> packet_msg)
+{
+  // Accumulate packets for recording only if someone is subscribed to the topic (for performance)
+  if (hw_interface_ &&
+      (packets_pub_->get_subscription_count() > 0 || packets_pub_->get_intra_process_subscription_count() > 0))
+  {
+    if (current_scan_msg_->packets.size() == 0)
+    {
+      current_scan_msg_->header.stamp = packet_msg->stamp;
+    }
+
+    pandar_msgs::msg::PandarPacket pandar_packet_msg{};
+    pandar_packet_msg.stamp = packet_msg->stamp;
+    pandar_packet_msg.size = packet_msg->data.size();
+    std::copy(packet_msg->data.begin(), packet_msg->data.end(), pandar_packet_msg.data.begin());
+    current_scan_msg_->packets.emplace_back(std::move(pandar_packet_msg));
+  }
+
+  std::tuple<nebula::drivers::NebulaPointCloudPtr, double> pointcloud_ts{};
+  nebula::drivers::NebulaPointCloudPtr pointcloud = nullptr;
+  {
+    std::lock_guard lock(mtx_driver_ptr_);
+    pointcloud_ts = driver_ptr_->ParseCloudPacket(packet_msg->data);
+    pointcloud = std::get<0>(pointcloud_ts);
+  }
+
+  if (pointcloud == nullptr)
+  {
+    // todo
```

Comment:
Is there a reason this is todo - can we at least send a warning for now?

Reply:
Ah, this has been addressed in [83a3e7a](https://github.com/tier4/nebula/pull/127/commits/83a3e7a6cbb42b71335fcf47797971cb5862707b) by introducing a watchdog timer. Even if this function here never runs at all (such as when no packets are coming in), the watchdog will print warnings if pointcloud deadlines are missed.

I will remove the `todo`.

Reply:
Addressed in [215d131](https://github.com/tier4/nebula/pull/127/commits/215d131e1fb28b350d177428f2011cd9b3879efa)